### PR TITLE
created Vision

### PR DESCRIPTION
**Pull request by AmitBrunicki**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import com.spikes2212.command.DashboardedSubsystem;
+import com.spikes2212.util.Limelight;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+public class Vision extends DashboardedSubsystem {
+
+    private static Vision instance;
+
+    public static final String CAMERA_NAME = "photoncamera";
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: PHOTON_VISION_CAMERA_NAME

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package frc.robot.subsystems;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import com.spikes2212.util.Limelight;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+public class Vision extends DashboardedSubsystem {
+
+    private static Vision instance;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: private static comes after public static

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    public static final String CAMERA_NAME = "photoncamera";
+
+    public static final int PIPELINE_CONE_INDEX = 1;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: index starts at 0

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    public static final int PIPELINE_CUBE_INDEX = 2;
+
+    public static final int PIPELINE_RETROREFLECTIVE_INDEX = 1;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: index starts at 0

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        return 0;
+    }
+
+    private double getLimelightYaw(){
+        return limelight.getHorizontalOffsetFromTargetInDegrees();
+    }
+
+    public boolean limelightIsOnTarget(){
+        return limelight.isOnTarget();
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: should use `limelight.hasTarget()`

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            return result.getBestTarget().getYaw();
+        }
+        return 0;
+    }
+
+    private double getLimelightYaw(){
+        return limelight.getHorizontalOffsetFromTargetInDegrees();
+    }
+
+    public boolean limelightIsOnTarget(){
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: `limelightHasTarget`

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        return limelight.isOnTarget();
+    }
+
+    public boolean photonvisionIsOnTarget(){
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: no need for this

File: vendordeps/photonlib.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            "artifactId": "PhotonLib-java",
+            "version": "v2023.2.1"
+        },
+        {
+            "groupId": "org.photonvision",
+            "artifactId": "PhotonTargeting-java",
+            "version": "v2023.2.1"
+        }
+    ]
+}
```
Comment (Spikes-2212) [inline vendordeps/photonlib.json]: add blank line at the end of the file

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    private static Vision instance;
+
+    public static final String CAMERA_NAME = "photoncamera";
+
+    public static final int PIPELINE_CONE_INDEX = 1;
+    public static final int PIPELINE_CUBE_INDEX = 2;
+
+    public static final int PIPELINE_RETROREFLECTIVE_INDEX = 1;
+    public static final int PIPELINE_APRILTAG_INDEX = 2;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: ditto

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import org.photonvision.targeting.PhotonPipelineResult;
+
+public class Vision extends DashboardedSubsystem {
+
+    private static Vision instance;
+
+    public static final String CAMERA_NAME = "photoncamera";
+
+    public static final int PIPELINE_CONE_INDEX = 1;
+    public static final int PIPELINE_CUBE_INDEX = 2;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java]: ditto

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Comment by AmitBrunicki** [comment]
now it is perfect

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
+package frc.robot.subsystems;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import com.spikes2212.util.Limelight;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+
+public class Vision extends DashboardedSubsystem {
+
+    public static final String PHOTON_VISION_CAMERA_NAME = "photoncamera";
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Vision.java:10]: should be `photonvision`

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]
Add a method that returns that april tag id from the limelight and a method that returns the pose3d from the limelight

File: src/main/java/frc/robot/subsystems/Vision.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    public static Vision getInstance(){
+        if (instance==null){
+            instance = new Vision(new PhotonCamera(PHOTON_VISION_CAMERA_NAME), new Limelight());
+        }
+        return instance;
+    }
+
+    private Vision(PhotonCamera photonCamera, Limelight limelight) {
+        super("vision");
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Vision.java:31]: get the namespaceName as a parameter and give it when you construct the instance

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**